### PR TITLE
Add GitHub Actions workflow

### PR DESCRIPTION
**Pull request by severinsimmler**
So this is basically the last thing we need to build and publish the wheels. I added a GitHub Actions workflow that runs every time something is pushed to the master branch. This is basically a matter of taste, we could also configure it to run [when someone pushes a specific tag](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-workflow-only-when-a-push-of-specific-tags-occurs) – I'm fine with everything.

If an event triggered the workflow, it will set up a Windows, macOS and Ubuntu machine, install everything, build wheels with `cibuildwheel` and upload both the wheels and source dist to PyPI using Poetry. This requires the project owner's PyPI username (`PYPI_USER`) and password (`PYPI_PW`) to be available as environment variables (you can set these in the [repository's settings page](https://github.com/lda-project/lda/settings/secrets/actions) as secrets). Please also see the GitHub docs on using secrets in GitHub Actions: https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions

The builds on all three platforms should work as expected. I tested it with my fork (and the `--dry-run` option for `poetry publish`): https://github.com/severinsimmler/lda/actions/runs/6866391327

Since the logs of the mentioned build are only temporarily available (a few weeks I think?), this would've been published:

**Linux**
```
lda-2.0.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
lda-2.0.0-cp39-cp39-musllinux_1_1_x86_64.whl
lda-2.0.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
lda-2.0.0-cp310-cp310-musllinux_1_1_x86_64.whl
lda-2.0.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
lda-2.0.0-cp311-cp311-musllinux_1_1_x86_64.whl
lda-2.0.0-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
lda-2.0.0-cp312-cp312-musllinux_1_1_x86_64.whl
```

**macOS**
```
lda-2.0.0-cp39-cp39-macosx_12_0_x86_64.whl
lda-2.0.0-cp310-cp310-macosx_12_0_x86_64.whl
lda-2.0.0-cp311-cp311-macosx_12_0_x86_64.whl
lda-2.0.0-cp312-cp312-macosx_12_0_x86_64.whl
```

**Windows**
```
lda-2.0.0-cp39-cp39-win_amd64.whl
lda-2.0.0-cp310-cp310-win_amd64.whl
lda-2.0.0-cp311-cp311-win_amd64.whl
lda-2.0.0-cp312-cp312-win_amd64.whl
```

**Source dist**
```
lda-2.0.0.tar.gz
```

Having all this, building and publishing wheels for e.g. Python 3.13 next year should be super easy and straightforward: we just have to add 3.13 to the pyproject.toml and trigger the GH Actions build.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
+name: CI & CD
+
+on:
+  push:
+    branches:
+      - master
+
+env:
+  PYTHONDONTWRITEBYTECODE: 1
```
Comment (severinsimmler) [inline .github/workflows/release.yml:9]: This avoids creating `.pyc` files which only makes sense on a local machine

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: actions/setup-python@v4
+        name: Install Python
+        with:
+          python-version: "3.12"
+
+      - name: Install Poetry
+        uses: snok/install-poetry@v1
```
Comment (ariddell) [inline .github/workflows/release.yml]: I'm sure snok is trustworthy but I'd rather not have this if it's not required. Can't one just ``pip install "poetry~=1.2"``? I do this in httpstan without issue, https://github.com/stan-dev/httpstan/blob/main/.github/workflows/main.yml

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
+name: CI & CD
+
+on:
+  push:
+    branches:
+      - master
```
Comment (ariddell) [inline .github/workflows/release.yml]: Is there a standard way of setting things up so it only publishes on tagged commits? Do we need separate github action workflows for testing and publishing? I haven't been using cibuildwheel so I'm not sure what the common practices are.

**Review by ariddell — CHANGES_REQUESTED** [review]
This is great. Thank you!

Would you mind removing all the old CI things? I think that will avoid confusing people. That is, remove:

- continuous_integration/
- appveyor.yml

**Comment by ariddell** [comment]
One more thing: shouldn't tests be run before publishing?

**Comment by ariddell** [comment]
If moving the tests around and switching over to pytest is the easiest way to do this, I'm completely fine with that.

Thanks again for getting the source install working. This alone is an enormous advance.

File: .github/workflows/release.yml (hunk 3 above)
Comment (severinsimmler) [reply .github/workflows/release.yml]: Yes, it's possible to publish only on tagged commits – will update it :+1: And we do not need a separate workflow for testing and publishing. This all happens inside of cibuildwheel and the tests run against every wheel (e.g. here for Python 3.11 and 3.12 on Linux):

image

The tests all passed in the [dry run of my fork](https://github.com/severinsimmler/lda/actions/runs/6866391327):

image

File: doc/source/whats_new.rst [hunk 4]
```diff
@@ -6,6 +6,11 @@
  What's New
 ============
 
+v3.0.0 (tba)
+============
+- Drop support for Python 3.6, 3.7 and 3.8
+- Wheels for Python 3.9, 3.10, 3.11 and 3.12
```
Comment (severinsimmler) [inline doc/source/whats_new.rst:12]: Dropping support for Python versions is a breaking change I guess and requires a major version bump?

Comment (severinsimmler) [reply doc/source/whats_new.rst:12]: (We may not forget the set the date here)

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
+name: CI & CD
+
+on:
+  push:
+    tags:
+      - '*.*.*'
```
Comment (severinsimmler) [inline .github/workflows/release.yml:6]: The build is triggered now when a tag is pushed

**Comment by severinsimmler** [comment]
@ariddell, I removed the unused files and the dry run looked good: https://github.com/severinsimmler/lda/actions/runs/6946738382/job/18898933693

As mentioned above, tests run inside of cibuildwheel. If one fails, the whole build fails.

I also tried to update the docs, but not 100% sure about everything.

**Comment by ariddell** [comment]
This all looks great. Curious about how cibuildwheel runs the tests -- but I can look at that later.

**Comment by severinsimmler** [comment]
Awesome, thanks! Not sure what cibuildwheel's default test command is (see also https://cibuildwheel.readthedocs.io/en/stable/options/#test-command), but we just do what's configured in our pyproject.toml:

https://github.com/lda-project/lda/blob/aee65d31606a6944818cda32d6f151f347989781/pyproject.toml#L67-L68

**Comment by ariddell** [comment]
@severinsimmler thank you so much for this work. If you would like credit as an author of the package, please add your name to the `.mailmap` file in a PR. This is how I've been keeping track of significant contributions.

**Pull request merged**